components/region_cache_memory_engine/src/region_label.rs

```diff
@@ -0,0 +1,419 @@
+        }
+    }
+    pub async fn watch_region_labels(&mut self) {
+        self.reload_all_region_labels().await;
+        'outer: loop {
+            let region_label_path = self.region_label_path();
+            let (mut stream, cancel) = stream::abortable(
+                self.meta_client.watch(
+                    Watch::of(region_label_path.clone())
+                        .prefixed()
+                        .from_rev(self.revision)
+                        .with_prev_kv(),
+                ),
+            );
+            info!("pd meta client creating watch stream"; "path" => region_label_path, "rev" => %self.revision);
+            while let Some(grpc_response) = stream.next().await {
+                match grpc_response {
+                    Ok(resp) => {
+                        self.revision = resp.get_header().get_revision();
+                        let events = resp.get_events();
+                        events.iter().for_each(|event| match event.get_type() {
+                            EventEventType::Put => {
+                                match serde_json::from_slice::<LabelRule>(
+                                    event.get_kv().get_value(),
+                                ) {
+                                    Ok(label_rule) => self.on_label_rule(&label_rule),
+                                    Err(e) => error!("parse put region label event failed"; "name" => ?event.get_kv().get_key(), "err" => ?e),
+                                }
+                            }
+                            EventEventType::Delete => {
+                                match serde_json::from_slice::<LabelRule>(
+                                    event.get_prev_kv().get_value()
+                                ) {
+                                    Ok(label_rule) => self.manager.remove_region_label(&label_rule.id),
+                                    Err(e) => error!("parse delete region label event failed"; "name" => ?event.get_kv().get_key(), "err" => ?e),
+                                }
+                            }
+                        });
+                    }
+                    Err(PdError::DataCompacted(msg)) => {
+                        error!("required revision has been compacted"; "err" => ?msg);
+                        self.reload_all_region_labels().await;
+                        cancel.abort();
+                        continue 'outer;
+                    }
+                    Err(err) => {
+                        error!("failed to watch region labels"; "err" => ?err);
+                        let _ = GLOBAL_TIMER_HANDLE
+                            .delay(std::time::Instant::now() + RETRY_INTERVAL)
+                            .compat()
+                            .await;
+                        cancel.abort();
+                        continue 'outer;
+                    }
+                }
+            }
+        }
+    }
+
+    async fn reload_all_region_labels(&mut self) {
```

Comment:
Should reload clear the current labels first?

Reply:
This will override them. I am matching the behavior with PD used in other cases.

Reply:
Note that this intends to run in a background thread, much like the resource manager does. Will start when server is started.